src/tsickle.ts

```diff
@@ -105,33 +105,61 @@ class ClosureRewriter extends Rewriter {
+   *    function statement; for overloads, name will have been merged.
+   */
+  emitFunctionType(fnDecls: ts.SignatureDeclaration[], extraTags: jsdoc.Tag[] = []): string[] {
     let typeChecker = this.program.getTypeChecker();
-    let sig = typeChecker.getSignatureFromDeclaration(fnDecl);
-
-    // Construct the JSDoc comment by reading the existing JSDoc, if
-    // any, and merging it with the known types of the function
-    // parameters and return type.
-    let jsDoc = this.getJSDoc(fnDecl) || [];
     let newDoc = extraTags;
+    let paramNamesToReturn: string[] = [];
+    const abstract = {tagName: 'abstract'};
+    let lens = fnDecls.map(fnDecl => fnDecl.parameters.length);
+    const minCount = Math.min(...lens);
+    const maxCount = Math.max(...lens);
+    const isOverloaded = fnDecls.length > 1;
+    const isConstructor = fnDecls[0].kind === ts.SyntaxKind.Constructor;
+    let paramData = {
+      names: new Array<Set<string>>(maxCount),
+      types: new Array<Set<string>>(maxCount),
+      returns: new Set<string>()
+    };
 
-    // Copy all the tags other than @param/@return into the new
-    // comment without any change; @param/@return are handled later.
-    for (let tag of jsDoc) {
-      if (tag.tagName === 'param' || tag.tagName === 'return') continue;
-      newDoc.push(tag);
-    }
-
-    // Abstract
-    if ((fnDecl.flags & ts.NodeFlags.Abstract)) {
-      newDoc.push({tagName: 'abstract'});
-    }
+    for (let fnDecl of fnDecls) {
+      // Construct the JSDoc comment by reading the existing JSDoc, if
+      // any, and merging it with the known types of the function
+      // parameters and return type.
+      let jsDoc = this.getJSDoc(fnDecl) || [];
+
+      // Copy all the tags other than @param/@return into the new
+      // comment without any change; @param/@return are handled later.
+      // TODO: handle for overloads. indexOf on the array doesn't match instances of the Tag type,
+      // so another matching strategy is needed to avoid dupes. Also, there may be problems if
+      // an annotation doesn't apply to all overloads.
+      for (let tag of jsDoc) {
+        if (tag.tagName === 'param' || tag.tagName === 'return' || isOverloaded) continue;
+        newDoc.push(tag);
+      }
 
-    // Parameters.
-    if (sig.parameters.length) {
+      // Abstract
+      if ((fnDecl.flags & ts.NodeFlags.Abstract) && (newDoc.indexOf(abstract) === -1)) {
+        newDoc.push(abstract);
+      }
+      // Merge the parameters into a single list of merged names and list of types
+      let sig = typeChecker.getSignatureFromDeclaration(fnDecl);
+      // Parameters.
       // Iterate through both the AST parameter list and the type's parameter
       // list, as some information is only available in the former.
-      for (let i = 0; i < sig.parameters.length; i++) {
+      for (let i = 0, l = sig.parameters.length; i < l; i++) {
```

Comment:
Oh, then this takes me back to my original comment -- we should use a profiler to guide performance rather than optimizing in anticipation.